### PR TITLE
py-avro: Update to version 1.10.0, improvements, and bugfixes

### PR DESCRIPTION
**Pull request by essandess**
py-avros: Update to version 1.10.0, improvements, and bugfixes

* Update to version 1.10.0
* Move to pypi source to avoud tar extraction issue on some systems
* Add tests

###### Type(s)

- [x] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 10.15.5 19F101
Xcode 11.5 11E608c 

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?

- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

File: python/py-avro/Portfile [hunk 1]
```diff
@@ -4,10 +4,11 @@ PortSystem          1.0
 PortGroup           github 1.0
 PortGroup           python 1.0
 
-github.setup        apache avro 1.9.1 release-
+github.setup        apache avro 1.10.0 release-
 revision            0
 name                py-${github.project}
 categories-append   devel
+fetch.type          git
```
Comment (reneeotten) [inline python/py-avro/Portfile:11]: why did you add this - it seem to work perfectly fine without it

File: python/py-avro/Portfile [hunk 2]
```diff
@@ -18,17 +19,36 @@ long_description    ${description}
...
+    post-destroot {
+        set docdir ${prefix}/share/doc/${subport}
+        xinstall -d ${destroot}${docdir}
+        xinstall -m 0644 -W ${worksrcpath} LICENSE.txt README.md \
+            ${destroot}${docdir}
+    }
 
     livecheck.type      none
+} else {
+    livecheck.type      pypi
```
Comment (reneeotten) [inline python/py-avro/Portfile:53]: you're using the `github` PG so please use the livecheck from GitHub as well. If you don't want pre-release versions set `github.livecheckregex` properly.

File: python/py-avro/Portfile [hunk 3]
```diff
@@ -18,17 +19,36 @@ long_description    ${description}
...
+                    size    1899682
 
 python.versions     37 38
 
 if {${name} ne ${subport}} {
     depends_build-append \
-                        port:py${python.version}-setuptools
-
-    worksrcdir          ${distname}/lang/py3
+                    port:py${python.version}-setuptools
```
Comment (reneeotten) [inline python/py-avro/Portfile:30]: the package doesn't build since you didn't include all dependencies - please start building things locally in trace-mode (using the `-v` flag) which will show you these things before submitting a PR

File: python/py-avro/Portfile [hunk 4]
```diff
@@ -18,17 +19,36 @@ long_description    ${description}
...
+                    sha256  4e94b43ebdb1081fd842644bda2caaddc586c1b608efe9236d348d01bf0866a4 \
+                    size    1899682
 
 python.versions     37 38
 
 if {${name} ne ${subport}} {
     depends_build-append \
-                        port:py${python.version}-setuptools
-
-    worksrcdir          ${distname}/lang/py3
```
Comment (reneeotten) [inline python/py-avro/Portfile:31]: I don't see the benefit of removing this and then needing to add build.dir, destroot.dir, and test.dir - if no obvious reason please revert

File: python/py-avro/Portfile (hunk 1 above)
Comment (essandess) [reply python/py-avro/Portfile:11]: Now for me. Here's what I see:

> `sudo port extract py37-avro`
```
:info:extract Executing:  cd "/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_python_py-avro/py37-avro/work" && /usr/bin/gzip -dc '/opt/local/var/macports/distfiles/py-avro/avro-1.9.1.tar.gz' | /usr/bin/tar -xf -
:debug:extract system:  cd "/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_python_py-avro/py37-avro/work" && /usr/bin/gzip -dc '/opt/local/var/macports/distfiles/py-avro/avro-1.9.1.tar.gz' | /usr/bin/tar -xf -
:info:extract apache-avro-8921826/lang/java/mapred/src/test/resources/org/apache/avro/mapreduce/mapreduce-test-input.avro/.part-r-00000.avro.crc: Failed to restore metadata
:info:extract tar: Error exit delayed from previous errors.
```

File: python/py-avro/Portfile (hunk 2 above)
Comment (essandess) [reply python/py-avro/Portfile:53]: Two points:
* pypi's automated version checking is *much* more reliable than many GitHub authors/repos in providing version tags, so it makes sense to me to use pypi for live checks even though the download source is from GitHub.
* If it's absolutely necessary to use GitHub version checking, please post the exact `Portfile` code needed.

File: python/py-avro/Portfile (hunk 4 above)
Comment (essandess) [reply python/py-avro/Portfile:31]: The original code now puts the *base* in `…/lang/py3`, but then fails because you have to go into another `lang/py3` below that to find `setup.py` to build.

File: python/py-avro/Portfile (hunk 3 above)
Comment (essandess) [reply python/py-avro/Portfile:30]: Show where? I don't see how this shows missing dependencies.

Here's what I see:

> `sudo port -v build py37-avro`
```
--->  Computing dependencies for py37-avro.
--->  Fetching distfiles for py37-avro
Cloning into '/opt/local/var/macports/build/_opt_local_ports_python_py-avro/py37-avro/work/avro-1.10.0'...
remote: Enumerating objects: 82, done.        
remote: Counting objects: 100% (82/82), done.        
remote: Compressing objects: 100% (77/77), done.        
remote: Total 50076 (delta 35), reused 38 (delta 4), pack-reused 49994        
Receiving objects: 100% (50076/50076), 16.87 MiB | 19.56 MiB/s, done.
Resolving deltas: 100% (26070/26070), done.
--->  Verifying checksums for py37-avro
--->  Extracting py37-avro
--->  Configuring py37-avro
--->  Building py37-avro
Executing:  cd "/opt/local/var/macports/build/_opt_local_ports_python_py-avro/py37-avro/work/avro-1.10.0/lang/py3" && /opt/local/Library/Frameworks/Python.framework/Versions/3.7/bin/python3.7 setup.py --no-user-cfg build -j12 
running build
running build_py
creating build
creating build/lib
creating build/lib/avro
copying avro/datafile.py -> build/lib/avro
copying avro/ipc.py -> build/lib/avro
copying avro/protocol.py -> build/lib/avro
copying avro/io.py -> build/lib/avro
copying avro/__init__.py -> build/lib/avro
copying avro/schemanormalization.py -> build/lib/avro
copying avro/tool.py -> build/lib/avro
copying avro/schema.py -> build/lib/avro
running egg_info
creating avro_python3.egg-info
writing avro_python3.egg-info/PKG-INFO
writing dependency_links to avro_python3.egg-info/dependency_links.txt
writing requirements to avro_python3.egg-info/requires.txt
writing top-level names to avro_python3.egg-info/top_level.txt
writing manifest file 'avro_python3.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'avro_python3.egg-info/SOURCES.txt'
copying avro/HandshakeRequest.avsc -> build/lib/avro
copying avro/HandshakeResponse.avsc -> build/lib/avro
copying avro/LICENSE -> build/lib/avro
copying avro/NOTICE -> build/lib/avro
copying avro/VERSION.txt -> build/lib/avro
creating build/lib/avro/tests
copying avro/tests/av_bench.py -> build/lib/avro/tests
copying avro/tests/gen_interop_data.py -> build/lib/avro/tests
copying avro/tests/run_tests.py -> build/lib/avro/tests
copying avro/tests/sample_http_client.py -> build/lib/avro/tests
copying avro/tests/sample_http_server.py -> build/lib/avro/tests
copying avro/tests/test_datafile.py -> build/lib/avro/tests
copying avro/tests/test_datafile_interop.py -> build/lib/avro/tests
copying avro/tests/test_enum.py -> build/lib/avro/tests
copying avro/tests/test_init.py -> build/lib/avro/tests
copying avro/tests/test_io.py -> build/lib/avro/tests
copying avro/tests/test_ipc.py -> build/lib/avro/tests
copying avro/tests/test_normalization.py -> build/lib/avro/tests
copying avro/tests/test_protocol.py -> build/lib/avro/tests
copying avro/tests/test_schema.py -> build/lib/avro/tests
copying avro/tests/test_script.py -> build/lib/avro/tests
running build_scripts
creating build/scripts-3.7
copying and adjusting scripts/avro -> build/scripts-3.7
changing mode of build/scripts-3.7/avro from 644 to 755
```

Comment (reneeotten) [reply python/py-avro/Portfile:30]: sorry, my bad to use trace-mode you'll have of course to use the `-t` flag to make sure it will only use dependencies that you actually specified.

Comment (essandess) [reply python/py-avro/Portfile:30]: Oh I see:
> `sudo port -vst build py37-avro`

File: python/py-avro/Portfile (hunk 4 above)
Comment (reneeotten) [reply python/py-avro/Portfile:31]: that's because you're using `fetch.type git` and I don't think you should

File: python/py-avro/Portfile (hunk 2 above)
Comment (reneeotten) [reply python/py-avro/Portfile:55]: you can easily find how to do this in other ports, but okay here it is: `github.livecheck.regex {([0-9.]+)}`

File: python/py-avro/Portfile (hunk 4 above)
Comment (essandess) [reply python/py-avro/Portfile:31]: The extract fails without it. See above.

> `sudo port extract py37-avro`
```
:info:extract Executing:  cd "/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_python_py-avro/py37-avro/work" && /usr/bin/gzip -dc '/opt/local/var/macports/distfiles/py-avro/avro-1.9.1.tar.gz' | /usr/bin/tar -xf -
:debug:extract system:  cd "/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_python_py-avro/py37-avro/work" && /usr/bin/gzip -dc '/opt/local/var/macports/distfiles/py-avro/avro-1.9.1.tar.gz' | /usr/bin/tar -xf -
:info:extract apache-avro-8921826/lang/java/mapred/src/test/resources/org/apache/avro/mapreduce/mapreduce-test-input.avro/.part-r-00000.avro.crc: Failed to restore metadata
:info:extract tar: Error exit delayed from previous errors.
```

File: python/py-avro/Portfile (hunk 3 above)
Comment (reneeotten) [reply python/py-avro/Portfile:32]: yes, as the PR checklist tells you to do...

File: python/py-avro/Portfile (hunk 1 above)
Comment (reneeotten) [reply python/py-avro/Portfile:11]: this suggests that you're using the old distfile (the name above say `avro-1.9.1.tar.gz`), clean the port first and try again after removing the `fetch.type` stuff and revert to the using `worksrcdir          ${distname}/lang/py3`

File: python/py-avro/Portfile (hunk 4 above)
Comment (reneeotten) [reply python/py-avro/Portfile:31]: see my comment above

File: python/py-avro/Portfile (hunk 1 above)
Comment (essandess) [reply python/py-avro/Portfile:11]: Nope:

> `sudo port clean --all py37-avro ; sudo port extract py37-avro`
```
:info:extract Executing:  cd "/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_python_py-avro/py37-avro/work" && /usr/bin/gzip -dc '/opt/local/var/macports/distfiles/py-avro/avro-1.9.1.tar.gz' | /usr/bin/tar -xf -
:debug:extract system:  cd "/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_python_py-avro/py37-avro/work" && /usr/bin/gzip -dc '/opt/local/var/macports/distfiles/py-avro/avro-1.9.1.tar.gz' | /usr/bin/tar -xf -
:info:extract apache-avro-8921826/lang/java/mapred/src/test/resources/org/apache/avro/mapreduce/mapreduce-test-input.avro/.part-r-00000.avro.crc: Failed to restore metadata
:info:extract tar: Error exit delayed from previous errors.
:info:extract Command failed:  cd "/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_python_py-avro/py37-avro/work" && /usr/bin/gzip -dc '/opt/local/var/macports/distfiles/py-avro/avro-1.9.1.tar.gz' | /usr/bin/tar -xf -
:info:extract Exit code: 1
:error:extract Failed to extract py37-avro: command execution failed
```

Comment (essandess) [reply python/py-avro/Portfile:11]: This fail is with the current Portfile, https://github.com/macports/macports-ports/blob/master/python/py-avro/Portfile

Comment (reneeotten) [reply python/py-avro/Portfile:11]: we don't really care what happens with the old Portfile, this PR is about how to update to the latest version. For that you can just update the `version`, `checksums`, and dependencies while keeping the `worksrcdir`.

Comment (essandess) [reply python/py-avro/Portfile:11]: > just update the `version`, `checksums`, and dependencies while keeping the `worksrcdir`.

@reneeotten That **will not work*** because of the extract issue described above. This is a bugfix. The only extract that I'm aware of that works in this repo is `fetch.type git`.

Comment (reneeotten) [reply python/py-avro/Portfile:11]: I don't know if you actually tried any of my suggestions, but the Portfile below works just fine:

```
> port -dv installed py38-avro
The following ports are currently installed:
  py38-avro @1.10.0_0 (active) platform='darwin 18' archs='x86_64' date='2020-07-08T16:03:53-0400'
```

```
# -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:ft=tcl:et:sw=4:ts=4:sts=4

PortSystem          1.0
PortGroup           github 1.0
PortGroup           python 1.0

github.setup        apache avro 1.10.0 release-
revision            0
name                py-${github.project}
categories-append   devel

platforms           darwin
license             Apache-2
maintainers         nomaintainer

description         Avro is a serialization and RPC framework.
long_description    ${description}

homepage            https://avro.apache.org/

checksums           rmd160  e398cc448a227de8754eb60848eb4a47f0633265 \
                    sha256  4e94b43ebdb1081fd842644bda2caaddc586c1b608efe9236d348d01bf0866a4 \
                    size    1899682

python.versions     37 38

if {${name} ne ${subport}} {
    depends_build-append \
                    port:py${python.version}-setuptools \
                    port:py${python.version}-codestyle \
                    port:py${python.version}-isort

    worksrcdir      ${worksrcpath}/lang/py3

    test.run        yes
    test.cmd        py.test-${python.branch}
    test.target
    test.env-append PYTHONPATH=${worksrcpath}/build/lib

    post-destroot {
        set docdir ${prefix}/share/doc/${subport}
        xinstall -d ${destroot}${docdir}
        xinstall -m 0644 -W ${worksrcpath}/../../ LICENSE.txt README.md \
            ${destroot}${docdir}
    }

    livecheck.type      none
} else {
    github.livecheck.regex {([0-9.]+)}
}
```

Comment (essandess) [reply python/py-avro/Portfile:11]: This does not work for me, same as all the comments above. There's an issue with the repo. Using `fetch.type git` fixes it.

> `sudo port clean --all py37-avro ; sudo port extract py37-avro`
```
:info:extract Executing:  cd "/opt/local/var/macports/build/_opt_local_ports_python_py-avro/py37-avro/work" && /usr/bin/gzip -dc '/opt/local/var/macports/distfiles/py-avro/avro-1.10.0.tar.gz' | /usr/bin/tar -xf -
:debug:extract system:  cd "/opt/local/var/macports/build/_opt_local_ports_python_py-avro/py37-avro/work" && /usr/bin/gzip -dc '/opt/local/var/macports/distfiles/py-avro/avro-1.10.0.tar.gz' | /usr/bin/tar -xf -
:info:extract apache-avro-f83cd83/lang/java/mapred/src/test/resources/org/apache/avro/mapreduce/mapreduce-test-input.avro/.part-r-00000.avro.crc: Failed to restore metadata
:info:extract tar: Error exit delayed from previous errors.
:info:extract Command failed:  cd "/opt/local/var/macports/build/_opt_local_ports_python_py-avro/py37-avro/work" && /usr/bin/gzip -dc '/opt/local/var/macports/distfiles/py-avro/avro-1.10.0.tar.gz' | /usr/bin/tar -xf -
:info:extract Exit code: 1
```

Comment (essandess) [reply python/py-avro/Portfile:11]: macOS 10.15.5 19F101
Xcode 11.5 11E608c

**Comment by ryandesign** [comment]
Why does this port fetch from github rather than from pypi like most other python module ports?

**Comment by essandess** [comment]
> Why does this port fetch from github rather than from pypi like most other python module ports?

A lot of python ports fetch from github. The original repos have additional stuff that pypi often doesn’t: tests, documentation, and the like. Personally, I always prefer to build from the original repo if possible.

**Comment by ryandesign** [comment]
Ok, if there are files in the github repo that we need that aren't available in the pypi distfile, that's a good reason to fetch from github. Otherwise, the default should be to fetch from pypi.

**Comment by reneeotten** [comment]
all files seem to be present in the PyPI tarball, except for the `README.md` but that file is pretty useless anyway. So yes, let's use PyPI that avoids having discussions about fetching from GitHub.

File: python/py-avro/Portfile (hunk 1 above)
Comment (essandess) [reply python/py-avro/Portfile]: Moved to pypi.

File: python/py-avro/Portfile (hunk 4 above)
Comment (essandess) [reply python/py-avro/Portfile:31]: Moved to pypi.

File: python/py-avro/Portfile [hunk 5]
```diff
@@ -1,13 +1,12 @@
...
-PortGroup           github 1.0
 PortGroup           python 1.0
 
-github.setup        apache avro 1.9.1 release-
+name                py-avro
+version             1.10.0
 revision            0
-name                py-${github.project}
-categories-append   devel
+categories-append   python devel
```
Comment (reneeotten) [inline python/py-avro/Portfile:9]: the `python` category is already added by the `python` PG so please don't append that again

File: python/py-avro/Portfile [hunk 6]
```diff
@@ -17,18 +16,27 @@ description         Avro is a serialization and RPC framework.
 long_description    ${description}
 
 homepage            https://avro.apache.org/
+master_sites        pypi:a/avro
+distname            avro-${version}
```
Comment (reneeotten) [inline python/py-avro/Portfile:20]: the `master_sites` and `distname` are the default set by the `python` PG, so please remove these redundant lines

File: python/py-avro/Portfile [hunk 7]
```diff
@@ -17,18 +16,27 @@ description         Avro is a serialization and RPC framework.
...
 
-    worksrcdir          ${distname}/lang/py3
+    test.run        yes
+    test.cmd        py.test-${python.branch}
+    test.target
+    test.env-append PYTHONPATH=${worksrcpath}/build/lib
 
     livecheck.type      none
+} else {
+    livecheck.type      pypi
```
Comment (reneeotten) [inline python/py-avro/Portfile:41]: this is also the default set by the `python` PG so please remove

File: python/py-avro/Portfile [hunk 8]
```diff
@@ -17,18 +16,27 @@ description         Avro is a serialization and RPC framework.
...
 
 if {${name} ne ${subport}} {
     depends_build-append \
-                        port:py${python.version}-setuptools
+                    port:py${python.version}-setuptools \
+                    port:py${python.version}-codestyle \
+                    port:py${python.version}-isort
 
-    worksrcdir          ${distname}/lang/py3
+    test.run        yes
```
Comment (reneeotten) [inline python/py-avro/Portfile:34]: you are missing a test-dependency on `py-pytest`

File: python/py-avro/Portfile (hunk 5 above)
Comment (essandess) [reply python/py-avro/Portfile:9]: Thanks done.

File: python/py-avro/Portfile (hunk 6 above)
Comment (essandess) [reply python/py-avro/Portfile:20]: Thanks done.

Would be good to find some cycles to update https://github.com/macports/macports-contrib/blob/master/pypi2port/pypi2port.py …

File: python/py-avro/Portfile (hunk 8 above)
Comment (essandess) [reply python/py-avro/Portfile:34]: Thanks done.

File: python/py-avro/Portfile (hunk 7 above)
Comment (essandess) [reply python/py-avro/Portfile:41]: Thanks done.

File: python/py-avro/Portfile (hunk 6 above)
Comment (reneeotten) [reply python/py-avro/Portfile]: that's true indeed... I personally don't use that, but use `upt`, which was part of the GSoC last year.

**Comment by reneeotten** [comment]
thanks @essandess
